File: github_cpp/3/64.cpp
#ifndef TREES_AVLTREE_H
#define TREES_AVLTREE_H

#include <iostream>
#include <list>
#include <vector>
#include <functional>

using namespace std;

class AVLTree {

private:
    
    struct Node {
        
        const int key;

        
        Node *previous = nullptr;

        
        Node *left = nullptr;

        
        Node *right = nullptr;

        
        int balance = 0;

        
        AVLTree *tree = nullptr;

        
        Node(const int k, AVLTree *t);

        
        Node(const int k, Node *p, Node *l, Node *r, AVLTree *t);

        
        ~Node();

        
        void upin(bool growedLeft);

        
        void upout(bool leftShrinked);

        
        void rightRotation();

        
        void leftRightRotation();

        
        void rightLeftRotation();

        
        void leftRotation();

        
        bool isRoot();

        
        bool isLeftFollower();

        
        int height();

        
        vector<int> *preorder() const;  

        
        vector<int> *inorder() const;   

        
        vector<int> *postorder() const; 
    };

    
    Node *root = nullptr;

public:
    
    ~AVLTree();

    
    bool search(const int) const;

    
    void insert(const int);

    
    void remove(const int);

    
    vector<int> *preorder() const;  

    
    vector<int> *inorder() const;   

    
    vector<int> *postorder() const; 


    friend ostream &operator<<(ostream &, const AVLTree &);
    friend class treeplotter;
    
    int getBalance(const int);

    
    Node* find(const int);

};

#endif 

File: github_code_index/java/2/64.java
/*
* AVL-Tree implementation
* srv_twry
* */

class Node{
    public static int data,height;
    Node left,right;

    public Node(int data){
        this.data=data;
        height=1;
        left=null;
        right=null;
    }
}

class AVLTree{

    public static Node root;

    public AVLTree(){
        root=null;
    }

    // A utility function to get height of the tree
    int height(Node N) {
        if (N == null)
            return 0;

        return N.height;
    }

    // A utility function to get maximum of two integers
    int max(int a, int b) {
        return (a > b) ? a : b;
    }



}